services/QuillLMS/app/models/cache_key.rb

```diff
@@ -0,0 +1,35 @@
+# Used to define common cache_keys in the app and force a common pattern
+module CacheKey
+
+  def self.all_classrooms(user, name:, groups: [])
```

Comment:
I think it's better to not have app-specific logic in this module, but rather use `model_key` from the app, and pass in all the necessary ingredients from the caller context.

Reply:
It's helpful in this RFC to see all of the proposed methods within CacheKey module, but I would agree with @happythenewsad, the methods should be in the caller context.

Reply:
I'm a bit torn.  Counter-point would be that it a central registry of cache-keys can help you avoid both double-naming things (and therefore having inaccurate "misses") when some other part of the code is already caching what you care about, and (more worryingly) collisions.

That said, if we use this pattern extensively, I could see this file having dozens of registered key patterns, at which point it's so unwieldy that it might not help solve the risk of collision anymore anyway.

Reply:
@happythenewsad @brendanshean @anathomical 

I think it would be helpful to explain my thinking on this class further. 

We technically don't need this `CacheKey` class, it's entirely an object lookup and pass-through methods to form an array for `Rails.cache.fetch`. You can use `Rails.cache.fetch` directly fine enough.

My concern is that when using `Rails.cache.fetch`, it's relatively complicated to know which model objects are usable for their `updated_at` timestamp as cache keys, e.g. you have to confirm that every child object is properly updating its parent objects (which may have edge cases), and you have to validate that an object's `updated_at` ensures everything attached is up to date. I'm concerned that this overhead will discourage developers from using caching (or encourage them to just copy/paste from an example somewhere else in the code base).

My thought with this class is that it might be easier to have an `"approved list"` of `cache_keys` that have been vetted/validated so that:

1.  Developers can use these approved keys with only a little knowledge of the caching setup, and just knowledge of "Is the data I want to cache `all teacher classrooms` level, `individual classroom` level, `classroom_unit` level, etc."
1. It lowers the barrier of entry for adding caching.
1. The cache keys follow a standard pattern

I'd be interested to hear feedback on this.

Reply:
Those are strong points.  I like the idea of the class being a place for "approved" cache_keys makes sense as this is a pattern we should be careful when using.   While I was worried about the file becoming too bloated, I think part of that was due to the generic name `CacheKey`.  What about a more specific name like: `LowLevelCacheKey`?  This would communicate the intent of the keys more effectively while preventing accidental inclusion of other types of cache keys.